---
title: C26447
ms.date: 03/22/2018
ms.topic: reference
f1_keywords: ["C26447"]
helpviewer_keywords: ["C26447"]
---
# C26447 DONT_THROW_IN_NOEXCEPT

The function is declared **`noexcept`** but calls a function that may throw exceptions.

C++ Core Guidelines:
[F.6: If your function may not throw, declare it noexcept](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f6-if-your-function-may-not-throw-declare-it-noexcept).

This rule amends another rule, [C26440 DECLARE_NOEXCEPT](c26440.md), which tries to find functions that are good candidates to be marked as **`noexcept`**. In this case, the idea is that once some function is marked as **`noexcept`**, it must keep its contract by not invoking other code that may throw exceptions.

- The Microsoft C++ compiler already handles straightforward violations like **`throw`** statements in the function body (see [C4297](/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4297)).
- The rule focuses only on function calls. It flags targets that are not **`constexpr`** and that can potentially throw exceptions; in other words they are not marked explicitly as non-throwing by using **`noexcept`**, **`__declspec(nothrow)`**, **throw()**.
- The compiler-generated target functions are skipped to reduce noise since exception specifications are not always provided by the compiler.
- The checker also skips special kinds of target functions that are expected to be implemented as **`noexcept`**; this rule is enforced by [C26439 SPECIAL_NOEXCEPT](c26439.md).

## Example
```cpp
#include <vector>
#include <string>
#include <istream>

std::vector<std::string> collect(std::istream& is) noexcept 
{
    std::vector<std::string> res;
    for (std::string s; is >> s;) // C26447, `operator bool()` can  throw, std::string's allocator can throw
        res.push_back(s);         // C26447, `push_back` can throw
    return res;
}
```
You can fix these warnings by removing noexcept from the function signature.
